Library/Source/NativeEngine.cpp

```diff
@@ -33,6 +33,37 @@ namespace Babylon
 {
     namespace
     {
```

Comment:
Definitely the weirdest "fix" in here, and the one I'm least sure about. bgfx doesn't seem to have a mechanism for asking it whether it's alive or not, but some JavaScript-owned objects (relevantly the `ProgramData`) can end up being destroyed after `NativeEngine` has already been destroyed and `bgfx::shutdown` has consequently already been called. They could also theoretically survive long enough to last into another bgfx initialization (if that's possible), which would be disastrous as their handles would now point to things inside bgfx that were really created and owned by somebody else. This global state holder was created to keep track of this. I'm not a fan of it, however, so if anybody has a recommendation of what would be better, please let me know.

Reply:
There are 2 resources to free here : the bgfx shader handle and the JS object. I'd try to keep a list of ProgramData (register in ctor, unregister in dtor). When bgfx::shutdown is called, I'd run thru all the registered ProgramData and free their bgfx handle. I'd then let JS engine free the ProgramData when ever it needs to. I think the same mechanism happens with TextureData. I prefer to be pro active in destruction than trying to determine a state(m_isBgfxInitialized/m_bgfxInitializationId) where I can or can't do something. The more states, the more potential bugs.

Reply:
Definitely agree that that's cleaner, and it seems the ticketed mechanism can be adapted for this purpose. I've refactored to try this other approach.